File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0
        r = 0
        freq = defaultdict(int)
        max_count=0
        while r<len(s):
            ch = s[r]
            freq[ch]+=1
            while freq[ch]>1:
                ch_l = s[l]
                freq[ch_l]-=1
                l+=1
            max_count = max(max_count, r-l+1)
            r+=1
        return max_count
                

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left=0;
        int right=0;
        int max = 0;
        Map<Character, Integer> freq= new HashMap<>();
        while(right<s.length()) {
            char ch = s.charAt(right);
            freq.put(ch, freq.getOrDefault(ch, 0)+1);
            while (freq.getOrDefault(ch,0)>1) {
                char ch_left = s.charAt(left);
                freq.put(ch_left, freq.get(ch_left)-1);
                left++;
            }
            max = Math.max(max, right - left + 1);
            right++;
 
        }
        return max;
    }
}